qiskit/circuit/quantumcircuitdata.py

```diff
@@ -26,6 +26,27 @@ class CircuitInstruction:
     """A single instruction in a :class:`.QuantumCircuit`, comprised of the :attr:`operation` and
     various operands.
 
+    .. note::
+
+        There is some possible confusion in the names of this class, :class:`~.circuit.Instruction`,
+        and :class:`~.circuit.Operation`, and this class's attribute :attr:`operation`.  Our
+        preferred terminology is by analogy to assembly languages, where an "instruction" is made up
+        of an "operation" and its "operands".
+
+        Historically, :class:`~.circuit.Instruction` came first, and originally contained the qubits
+        it operated on and any parameters, so it was a true "instruction".  Over time,
+        :class:`.QuantumCircuit` became responsible for tracking qubits and clbits, and the class
+        became better described as an "operation".  Changing the name of such a core object would be
+        a very unpleasant API break for users, and so we have stuck with it.
+
+        This class was created to provide a formal "instruction" context object in
+        :class:`.QuantumCircuit.data`, which had long been made of ad-hoc tuples.  With this, and
+        the advent of the :class:`~.circuit.Operation` interface for adding more complex objects to
+        circuits, we took the opportunity to correct the historical naming.  For the time being,
+        this leads to an awkward case where :attr:`.CircuitInstruction.operation` is often an
+        :class:`~.circuit.Instruction` instance, but as the :class:`.Operation` interface gains more
+        use, this confusion will hopefully abate.
```

Comment:
Could you provide more commentary on the relationship between `Operation` and `Instruction`? Is `Operation` intended to replace `Instruction`? Will `Instruction` be deprecated?

Reply:
I can say that `Instruction` implements the `Operation` interface (and I can write that in the `Instruction` documentation too, if you think that's helpful?).  I'm not keen to talk about any potential future deprecation of `Instruction` in public documentation, because that isn't 100% decided yet - it's currently a bit too entwined in everything (the class has been in since the very first commit of Terra!) for us to know for certain if there's a killer reason we can't _completely_ remove it.

Reply:
> I can say that Instruction implements the Operation interface

Sure, this would be helpful. Though I think "abstract base class" would be clearer than "interface" here and elsewhere.

Reply:
Done (very belatedly) in a7ce93e4b.